src/debug.rs

```diff
@@ -1,129 +1,248 @@
-use crate::{PluginState, RayCastSource};
-use bevy::prelude::*;
+#![cfg(feature = "debug")]
+
 use std::marker::PhantomData;
 
-pub struct DebugCursor<T> {
-    _phantom: PhantomData<T>,
+use bevy::ecs::schedule::ShouldRun;
+use bevy::prelude::*;
+use bevy::utils::HashMap;
+
+use crate::{Debug, Enabled, PluginState, RayCastSource};
+
+pub struct DebugResource {
+    pub cube_size: f32,
+    pub cube_tail_scale: f32,
+    pub material: Option<Handle<StandardMaterial>>,
+    pub tip_mesh: Option<Handle<Mesh>>,
+    pub tail_mesh: Option<Handle<Mesh>>,
 }
-impl<T> Default for DebugCursor<T> {
+
+impl Default for DebugResource {
     fn default() -> Self {
-        DebugCursor {
-            _phantom: PhantomData::default(),
+        Self {
+            cube_size: 0.04,
+            cube_tail_scale: 20.0,
+            material: None,
+            tip_mesh: None,
+            tail_mesh: None,
         }
     }
 }
 
-pub struct DebugCursorTail<T> {
-    _phantom: PhantomData<T>,
-}
-impl<T> Default for DebugCursorTail<T> {
+pub(crate) struct SourceToCursorMap<T>(HashMap<Entity, Entity>, PhantomData<T>);
+
+impl<T> Default for SourceToCursorMap<T> {
     fn default() -> Self {
-        DebugCursorTail {
-            _phantom: PhantomData::default(),
-        }
+        Self(Default::default(), PhantomData)
     }
 }
 
-pub struct DebugCursorMesh<T> {
+pub struct DebugCursor<T> {
     _phantom: PhantomData<T>,
 }
-impl<T> Default for DebugCursorMesh<T> {
+
+impl<T> Default for DebugCursor<T> {
     fn default() -> Self {
-        DebugCursorMesh {
+        DebugCursor {
             _phantom: PhantomData::default(),
         }
     }
 }
 
-/// Updates the 3d cursor to be in the pointed world coordinates
-#[allow(clippy::type_complexity)]
-#[allow(clippy::too_many_arguments)]
-pub fn update_debug_cursor<T: 'static + Send + Sync>(
-    mut commands: Commands,
-    state: Res<PluginState<T>>,
+fn add_debug_cursor<T: 'static + Send + Sync>(
+    commands: &mut Commands,
+    source_entity: Entity,
+    mut source: &mut RayCastSource<T>,
+    sources_to_cursors: &mut ResMut<SourceToCursorMap<T>>,
```

Comment:
Do we need to maintain a separate resource that tracks entity pairs? Wouldn't it make more sense to add a component onto the `RayCastSource` itself? In that case there is no need to maintain a map, instead the `DebugCursor` just looks at its parent (a `RayCastSource`) to know where it needs to go.

While I don't think the current implementation in `master` is ideal, especially because it just sticks the debug cursor mesh right in the Camera entity, it's pretty simple to reason about, which I really value for this library.

I could be convinced your implementation is the right choice, I'm just wary of adding complexity unless we really need it.

Reply:
Sorry for the late response, I've been on vacation.

While I agree with you on the account of simplicity I cannot find a simple way to query the parent/child relationship with a specific component. That means that I would have to iterate over all child entities of the RayCastSource (which could be the camera/player with many children) and manually check, if this entity has the DebugCursor component.
As the debugging functionality is likely to be used in debugging use cases, the performance of the already slow debug build (in general) would get even worse.

IMO the gained performance outweighs the complexity in this case, but if you think the impact could be tollerated in these cases, I can modify it accordingly of course.

Reply:
>Sorry for the late response, I've been on vacation.

No worries 🙂.

> I cannot find a simple way to query the parent/child relationship with a specific component. [...]

I think you can do this a few ways, one way I might approach it:

```rust
// ...
parent_query: Query<RayCastSource>,
child_query: Query<DebugCursor>
// ...

for parent in parent_query.iter() {
    // Use the debug entity stored in the parent to get that DebugCursor entity
    child_query.get(parent.debug_entity)
}
```